### DISCUSSION TITLE
Polymorphism without discriminator column

### DISCUSSION THREAD

**Discussion by santalvarez** [sqlalchemy/sqlalchemy#10063] [Usage Questions] [answered]
Hi, Im trying to implement polymorphism in one of my classes but i need to do it without a discriminator column. 
The thing is I'm getting this error

```
Attempting to flush an item of type <class 'app.models.inventory.components.host.host_provider1.HostProvider1'> as a member of collection "OperatingSystem.hosts". 
Expected an object of type <class 'app.models.inventory.components.host.base.HostBase'> or a polymorphic subclass of this type. 
If <class 'app.models.inventory.components.host.host_provider1.HostProvider1'> is a subclass of <class 'app.models.inventory.components.host.base.HostBase'>, configure mapper "mapped class HostBase->host" to load this subtype polymorphically, or set enable_typechecks=False to allow any subtype to be accepted for flush. 
```

Here are my classes:

```python
class OperatingSystem(db.Model, Component):
    __tablename__ = "operating_system"
    __table_args__ = (UniqueConstraint("name", "version", "patch", name="os_uc"),)

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    name = Column(Enum(OperatingSystemName), nullable=False)
    version = Column(String(30), nullable=False)
    patch = Column(String(30), default="", nullable=False)

    # Many Hosts to one OS
    hosts = relationship("HostBase", back_populates="os", cascade_backrefs=False)

class HostBase(db.Model, Component):
    """ Contains the common fields for all Host objects.
    Other Host<Provider> classes will contain only the fields that are
    specific to that provider.
    """
    __tablename__ = "host"

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    device_id = Column(BigInteger, ForeignKey("device.id"), unique=True, nullable=False)
    os_id = Column(BigInteger, ForeignKey("operating_system.id"))

    # Many Hosts to one OS
    os = relationship("OperatingSystem", back_populates="hosts", lazy="joined")
    device = relationship("Device", back_populates="host", uselist=False)

class HostProvider1(HostBase):
    __mapper_args__ = {
        "polymorphic_load": "inline",
    }
    @declared_attr
    def ip_local(cls):
        return HostBase.__table__.c.get("ip_local", Column(String(20), nullable=True))

    @declared_attr
    def mac_address(cls):
        return HostBase.__table__.c.get("mac_address", Column(String(50), nullable=True))

```

Basically i have a table (host) where im storing information about different computers. I receive this information from different providers and not all providers have every field in the host table. So my goal is that when i get info from a provider, lets say provider1 I go and make a select to get the existing HostProvider1 and then I compare it with the incoming HostProvider1 and update the fields that are common to them only.

I've looked online and some say to add a discriminator column but it just won't work in this case given that the same row can be used by multiple host providers. What can I do?

**Comment by CaselIT** [comment]
Hi

>  just won't work in this case given that the same row can be used by multiple host providers

I don't think polymorphism is the right solution here, since that's used when a row is of a particular type only. You cannot have a row that both one concrete type and another at the same type

**Reply by santalvarez** [reply]
Ok, what do you think might be the way to solve this? Have you ever seen someone with the same issue like mine?

**Reply by CaselIT** [reply]
> Basically i have a table (host) where im storing information about different computers. I receive this information from different providers and not all providers have every field in the host table. So my goal is that when i get info from a provider, lets say provider1 I go and make a select to get the existing HostProvider1 and then I compare it with the incoming HostProvider1 and update the fields that are common to them only.

can you better explain what you are trying to do? From here it just seems you have a table with nullable columns, you select a row using the primary column and compare the non nullable values with the ones you have in input?

**Reply by gordthompson** [reply]
I tend to agree with @CaselIT that polymorphism is not really applicable here. If a single `host` row can apply to multiple providers then it is not really a "provider_1" row or a "provider_2" row. You can use views to *display* the row in "provider_1" or "provider_2" format, but that's a presentation issue, not a persistence issue.

Also, what is a `host` row really supposed to represent? If two different providers have "Dell OptiPlex 5000" servers then they have the same computer *model* but not the same *device*. The actual devices would be identified by something like a combination of make ("Dell"), model ("OptiPlex 5000"), and serial number.

**Reply by santalvarez** [reply]
Ok I'll explain it in more detail.

A host contains the information about the operating system of a computer. It has fields like local and vpn ip addresses, mac address, the version of the OS and other more custom fields like "is_antivirus_installed". I have different providers that can give me this information and I need to merge it and centralize it in one place.

The thing is not all providers will support the same fields so lets say **provider1** populates the field "is_antivirus_installed" on the host table and later I receive information from **provider2** that does not have that field I don't want to replace the "is_antivirus_installed" field with **Null**. So know you may thing well if the incoming field is **Null** then don't update it but that field may be null intentionally and in that case I will want to set the db field to null (think of a vpn ip that may be null when the user of that device is not connected).

This is the reason I thought polymorphism would be perfect because, in theory, it solved my problem.

Also the hosts are related to a Device table that has a serial_number attribute that is the primary business key.

I hope I'm making myself clear :)

**Reply by gordthompson** [reply]
If a provider does not supply a particular value like `vpn_ip` then you can construct your input data such that `vpn_id` is simply omitted from the list of attributes to update. OTOH, if the provider does supply a `vpn_ip` you can include it and it will get updated, even if it is None.

```python
class Base(DeclarativeBase):
    pass


class Host(Base):
    __tablename__ = "host"
    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=False)
    name: Mapped[str] = mapped_column(String(50))
    vpn_ip: Mapped[str] = mapped_column(String(50), nullable=True)


Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

engine.echo = True

# initial data
with Session(engine) as sess, sess.begin():
    sess.add(Host(id=1, name="original_1", vpn_ip="1.1.1.1"))

def update_host(host: Host, new_data: dict):
    for k, v in new_data.items():
        host.__setattr__(k, v)


with Session(engine) as sess, sess.begin():
    host_1 = sess.get(Host, 1)
    update_host(host_1, dict(name="new_1"))
"""
UPDATE host SET name=? WHERE host.id = ?
[generated in 0.00044s] ('new_1', 1)
COMMIT
"""

with Session(engine) as sess, sess.begin():
    host_1 = sess.get(Host, 1)
    update_host(host_1, dict(name="newer_1", vpn_ip=None))
"""
UPDATE host SET name=?, vpn_ip=? WHERE host.id = ?
[generated in 0.00041s] ('newer_1', None, 1)
COMMIT
"""
```

**Reply by santalvarez** [reply]
Gord, thanks for the reply and example.

The way I have my application set up right now is that i receive the information from providers (obviously each one comes in their own format) and format it and build a Host object. So then I can do something like this:

```python
class Host(Base):
    __tablename__ = "host"
    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=False)
    name: Mapped[str] = mapped_column(String(50))
    vpn_ip: Mapped[str] = mapped_column(String(50), nullable=True)

    def update(self, new: Host):
        self.name = name
        self.vpn_ip = vpn_ip

# This will be selected from db
old_host = Host()
# This will be contructed from the provider info I received
new_host = Host()

old_host.update(new_host)
```

In my polymorphism setup, HostProvider1's update method would look like this:

```python
def update(self, new):
    self.ip_local = new.ip_local
    self.mac_address = new.mac_address

    # This is the update method of HostBase
    super().update(new)
```

The project Im working on is quite big so I prefer to work with objects instead of dicts as it makes it easier to maintain in the future if we add or change fields in the table.
We are using version 1.4 btw

**Reply by gordthompson** [reply]
You can have a base class and provider-specific subclasses, just don't bother with polymorphism:

```python
class HostBase(ModelBase):
    __tablename__ = "host"
    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=False)
    name: Mapped[str] = mapped_column(String(50))


class Provider1Host(HostBase):
    vpn_ip: Mapped[str] = mapped_column(String(50), nullable=True)


class Provider2Host(HostBase):
    pass  # no additional attributes
```

Then, when importing from "provider_1" you can create a `Provider1Host` object, merge it into the session, and persist it

```python
with Session(engine) as sess, sess.begin():
    host_for_provider_1 = Provider1Host()
    host_for_provider_1.id = 1
    host_for_provider_1.name = "newer_1"
    host_for_provider_1.vpn_ip = None
    sess.merge(host_for_provider_1)
"""
UPDATE host SET name=?, vpn_ip=? WHERE host.id = ?
[generated in 0.00039s] ('newer_1', None, 1)
COMMIT
"""
```

**Reply by santalvarez** [reply]
I removed the __mapper_args__ that I had in my provider classes but the problem persists.

The error ocurrs when updating the **os** relationship, given that it is defined in **HostBase** and not in the provider host classes.
```
Expected an object of type <class 'app.models.inventory.components.host.base.HostBase'> or a polymorphic subclass of this type.
```

Should a relationship in HostBase work with its subclasses?

Here is a snippet that demonstrates the error:

```python
from __future__ import annotations
from sqlalchemy.engine.create import create_engine
from sqlalchemy.ext.declarative import declared_attr

from sqlalchemy.orm import (
    declarative_base,
    relationship,
    sessionmaker,
)
from sqlalchemy import (
    Column,
    ForeignKey,
    UniqueConstraint,
    BigInteger,
    String,
)

Base = declarative_base()

class HostBase(Base):
    __tablename__ = "host"

    id = Column(BigInteger, primary_key=True)
    os_id = Column(BigInteger, ForeignKey("operating_system.id"))

    # Many Hosts to one OS
    os = relationship("OperatingSystem", back_populates="hosts", lazy="joined")

    def update(self, new):
        if new.os:
            self.os = new.os

# pylint: disable=no-self-argument
class HostProvider1(HostBase):
    @declared_attr
    def ip_local(cls):
        return HostBase.__table__.c.get("ip_local", Column(String(20), nullable=True))

    @declared_attr
    def mac_address(cls):
        return HostBase.__table__.c.get("mac_address", Column(String(50), nullable=True))


    def update(self, new):
        self.ip_local = new.ip_local
        self.mac_address = new.mac_address

        super().update(new)


class OperatingSystem(Base):
    __tablename__ = "operating_system"
    __table_args__ = (UniqueConstraint("name", "version", "patch", name="os_uc"),)

    id = Column(BigInteger, primary_key=True)
    name = Column(String(49), nullable=False)
    version = Column(String(30), nullable=False)
    patch = Column(String(30), default="", nullable=False)

    # One OS to many Hosts
    hosts = relationship("HostBase", back_populates="os")

# change to your connection string, pg it's just an example
e = create_engine("sqlite://", echo=True)


Base.metadata.create_all(e)
session_factory = sessionmaker(e)
s = session_factory()
host = HostProvider1(id=1, ip_local="192.168.0.1", mac_address="as-as-as-as",
                      os=OperatingSystem(id=1, name="foo", version="1", patch=""))
s.add(host)
s.commit()

os = OperatingSystem(id=2, name="foo", version="2", patch="")
try:
    host.os = os
finally:
    Base.metadata.drop_all(e)
```

**Reply by CaselIT** [reply, marked as answer]
In this use case I would use different classes that aren't mapped at all, that contain the information obtained from a provided and know how to update a particular mapped class.

Something like this (note that it's mostly pseudocode)
```py
class DbClass(Base):
  __tablename__='something'
  sn = Column(Text(), primary_key=True)
  name = Column(String(49), nullable=False)
  version = Column(String(30), nullable=False)
  patch = Column(String(30), default="", nullable=False)
  ...

@dataclass
class Host1:
  sn: str
  name: str

  def update_model(self, model):
    assert model.sn == sn
    model.name = self.name

@dataclass
class Host2:
  sn: str
  version: str

  def update_model(self, model):
    assert model.sn == sn
    model.name = self.name


def load_host(data):
  # load the proper host instance
  return loaded_host

def update_model(session, host_info):
  host = session.get(DbClass, host_info.sn)
  # this method also creates missing rows in the db
  if host is None:
    host = DbClass(sn=host_info.sn)
    session.add(host)
  host_info.update_model(host)
  session.flush()

def main():
  with Session(...) as session:
    host_info = load_host(...)
    update_model(session, host_info)
```

**Reply by gordthompson** [reply]
@CaselIT has a better approach than mine. Further testing of my suggestion showed that things got awkward if more than one ORM subclass tried to use the same attribute.

**Reply by santalvarez** [reply]
Thanks a lot for the replies! The solution posted by Federico works great! 👍
